packages/openapi-typescript/src/transform/schema-object.ts

```diff
@@ -163,31 +199,25 @@ export function transformSchemaObjectWithComposition(
 
   // core + allOf: intersect
   const coreObjectType = transformSchemaObjectCore(schemaObject, options);
-  const allOfType = collectCompositions(schemaObject.allOf ?? []);
+  const allOfType = collectCompositions(
+    schemaObject.allOf ?? [],
+    schemaObject.required,
+  );
   if (coreObjectType || allOfType.length) {
-    let allOf: ts.TypeNode | undefined = allOfType.length
+    const allOf: ts.TypeNode | undefined = allOfType.length
       ? tsIntersection(allOfType)
       : undefined;
-    // add required props
-    if (
-      allOf &&
-      "required" in schemaObject &&
-      Array.isArray(schemaObject.required)
-    ) {
-      allOf = tsWithRequired(
```

Comment:
Change: don’t indiscriminately add `WithRequired<…>`; instead, only add it when it’s actually needed, and with the proper keys